CT-1808 Bugfix for users with multiple roles

stephenrichards opened the pull request:
A bug was reported where users with multiple roles (manager, responder and approver)
were unable to mark a case as responded.

This is caused by the Case::Base#team_for_user method which was used to determine
which team to pass into the state machine as the acting team.  It just took the
first team that was common to both the case (i.e. on the assignements) and the User#team_roles.
This took no account of the role of the team, therefore we had the clash of StateMachine
events being called with a team which had no rights to perform that event.

This has been replaced by two methods:

 - Case::Base#team_for_assigned_user(user, role)
 - Case::Base#team_for_unassigned_user(user, role)

Note that both methods now take role as a parameter, to ensure that the correct matching team
is returned.

the difference between the two methods are that #team_for_assigned_user looks for an intersect between
the user's teams, and the accepted assignments for that user on the case, whereas #team_for_unassigned_user
looks for an intersect between the user's teams and teams on assignments in  unassigned or pending (not necessarily
for the same user).

misaka commented on app/mailers/assignment_mailer.rb:
```diff
@@ -7,23 +7,48 @@ def new_assignment(assignment, recipient)
     mail(to: recipient)
   end
 
+  def ready_for_approver_review(assignment)
```
This class, `AssignmentMailer`, doesn't feel like the right place to put this. This mail isn't really about assignments ... I think we should move this into another file, although I'm not sure how best to organise these mailers. Maybe we should have a `ActionNotificationsMailer` which are all about notifying the user about actions to be taken within the system? Versus a mailer that may be a digest of the days activities ...

misaka commented on app/mailers/assignment_mailer.rb:
```diff
@@ -7,23 +7,48 @@ def new_assignment(assignment, recipient)
+  def ready_for_approver_review(assignment)
+    RavenContextProvider.set_context
+
+    @assignment = assignment
```
We don't expose a view, we just pass in values to GovUK Notify, so we don't need to use `@assignment`, just plain `assignment is enough`.

aliuk2012 replied:
Good point 👍  and fixed

misaka commented on app/services/case_approval_service.rb:
```diff
@@ -60,5 +66,18 @@ def bypass_approval_for_team(team)
+                       .for_team(current_info.team)
+                       .first
+
+      AssignmentMailer
```
Why pull out `assignment`? It's not obvious why it's needed and I think it'd read better if you either just passed in `kase` and the mailer sorts out the approver to notify, or this service sorts out the approver and passes it through.

This goes along with changing the mailer to no just be about assignments.

aliuk2012 replied:
So I did try to pass in just a case but the thing is we then have to lookup the approvers details for the mailer.  I also came across some weird errors relating to case being decorated and Stephen said that I should use kase.object when calling ready_for_approver_review but then I ran into other errors of 'NoMethodfound 'object' is undefined/

misaka replied:
What about doing this:

```
AssignmentMailer.ready_for_approver_review(assignment.user, assignment.case).deliver_later
```

Again, it's just that it's not obvious why `ready_for_approver_review` is getting an assignment object, splitting it out like this is making intentions much clearer.

misaka commented on app/services/case_approval_service.rb:
```diff
@@ -34,16 +34,22 @@ def process_approval
           approve_and_progress_as_normal(assignment)
         end
+
+        assignment.reload
```
Still needed? It's not obvious what bit of the code needs this to be reloaded.

aliuk2012 replied:
the assignment.kase.current_state is not updated unless assignment is reloaded. I need the case current state to be correct for the notify private method to work as expected. I agree it is not clear as to what or why this code is used for. For the time being I'll move it down to the notify_next_approver private method and run @kase.reload.

misaka commented on app/services/case_approval_service.rb:
```diff
@@ -34,16 +34,22 @@ def process_approval
       @result = :error
     end
+
+    notify_press_or_private if @result == :ok
```
I know that this service already does this, but I don't think it should be talking about `press` or `private` ... it should be `next approver`, or something more generic, so that it'll make sense with other case types.

aliuk2012 replied:
Good point 👍  and fixed

misaka commented on app/services/response_uploader_service.rb:
```diff
@@ -30,6 +31,7 @@ def upload!
       attachments = @uploader.process_files(@uploaded_files, @type)
       transition_state(attachments)
       @result = :ok
+      notify_next_approver if @result == :ok && @action == 'upload-approve'
```
`@result` isn't ever going to be not `:ok` here ... it was just assigned. :)

misaka replied:
Also, I liked the pattern in the `CaseApprovalService` where this notification is sent outside of rescue block so that it doesn't have an effect on the result of the action.

aliuk2012 replied:
Sure, I agree I like the pattern in the ApprovalService but if I change it this there it has other impacts like how the controllers action processes the returned values and then theres all the tests so I dont think. its worth all the additional work for this ticket. We should standardise our approach and I think a separate ticket for it would be better suited

misaka replied:
I don't see why this would be a big change ... all it would mean doing is moving the code in this method into a begin block, and then performing the `notify_next_approver` step outside of that block, leaving the `if @result == :ok && ...` check that's already here.

Although I agree that if it's a bigger change, this isn't the time to do it here,

misaka commented on app/services/response_uploader_service.rb:
```diff
@@ -102,4 +104,19 @@ def approve_and_progress_as_normal(filenames)
 
+  def notify_next_approver
+    if @case.current_state
+           .in?(%w( pending_press_office_clearance pending_private_office_clearance ))
```
This is just begging to cause a problem where we change this list of states in one location, either here or `CaseApprovalService`, and forget to change the other. Is there an easy way to centralise this that makes sense?

aliuk2012 replied:
While its already caused a problem and delay with getting this feature out by having two services "clearing" a case.

misaka commented on app/services/response_uploader_service.rb:
```diff
@@ -25,20 +25,26 @@ def seed!(filepath)
+        @attachments = @uploader.process_files(@uploaded_files, @type)
+        transition_state(@attachments)
+        @result = :ok
+        @attachments
```
Unnecessary.

stephenrichards committed:
CT-1808 Bugfix for users with multiple roles

A bug was reported where users with multiple roles (manager, responder and approver)
were unable to mark a case as responded.

This is caused by the Case::Base#team_for_user method which was used to determine
which team to pass into the state machine as the acting team.  It just took the
first team that was common to both the case (i.e. on the assignements) and the User#team_roles.
This took no account of the role of the team, therefore we had the clash of StateMachine
events being called with a team which had no rights to perform that event.

This has been replaced by two methods:

 - Case::Base#team_for_assigned_user(user, role)
 - Case::Base#team_for_unassigned_user(user, role)

Note that both methods now take role as a parameter, to ensure that the correct matching team
is returned.

the difference between the two methods are that #team_for_assigned_user looks for an intersect between
the user's teams, and the accepted assignments for that user on the case, whereas #team_for_unassigned_user
looks for an intersect between the user's teams and teams on assignments in  unassigned or pending (not necessarily
for the same user).
